spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala

```diff
@@ -451,6 +468,23 @@ class CometSparkSessionExtensions
       }
     }
   }
+
+  // CometExec already wraps a `ColumnarToRowExec` for row-based operators. Therefore,
+  // `ColumnarToRowExec` is redundant and can be eliminated.
+  //
+  // It was added during ApplyColumnarRulesAndInsertTransitions' insertTransitions phase when Spark
+  // requests row-based output such as `collect` call. It's correct to add a redundant
+  // `ColumnarToRowExec` for `CometExec`. However, for certain operators such as
+  // `CometCollectLimitExec` which overrides `executeCollect`, the redundant `ColumnarToRowExec`
+  // makes the override ineffective. The purpose of this rule is to eliminate the redundant
+  // `ColumnarToRowExec` for such operators.
+  case class EliminateRedundantColumnarToRow(session: SparkSession) extends Rule[SparkPlan] {
```

Comment:
Hmm I'm trying to understand why this is necessary. The test passes even if I remove this rule.

Reply:
I see, so if we have the extra `ColumnarToRowExec`, the code will go through its `executeCollect` instead which will call `doExecuteÇolumnar`, instead of calling the `executeCollect` in the `CollectLimitExec` itself. 

I think we can probably do the same for `CometTakeOrderedAndProjectExec` too - Spark has an `executeCollect` implementation for this too. However, I don't know how useful it is since `executeCollect` is not often used? cc @viirya

Reply:
Thanks, makes sense. This also removes `ColumnarToRowExec` from the plan even if instead of `executeCollect`, `doExecute` is used, but I think it is OK since `doExecute` itself calls `ColumnarToRowExec`